src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,53 +1,121 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+
 /**
  * <p>Реалізувати свій ArrayList який імплементує інтерфейс List. Дотриматися основних вимог щодо
  * реалізації ArrayList (default capacity, newCapacity...)</p>
  */
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private Object[] elementData;
+    private int size;
+
+    public ArrayList() {
+        elementData = new Object[DEFAULT_CAPACITY];
+        size = 0;
+    }
 
     @Override
     public void add(T value) {
-
+        if (!ensureCapacity()) {
+            elementData = grow(size);
+        }
+        elementData[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        checkCapacity();
+        if (index < 0 || index > size) {
+            throw new ArrayIndexOutOfBoundsException("Incorrect index");
+        }
+        if (ensureCapacity()) {
+            System.arraycopy(elementData, index, elementData, index + 1, size - index);
+            elementData[index] = value;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        if (size + list.size() > elementData.length) {
+            elementData = grow(size + list.size());
+        }
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return (T) elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndex(index);
+        elementData[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < 0 || index > size) {
+            throw new ArrayIndexOutOfBoundsException("Incorrect index");
+        }
+        T removedValue = (T) elementData[index];
+        int newSize;
+        if ((newSize = size - 1) > index) {
+            System.arraycopy(elementData, index + 1,
+                    elementData, index, newSize - index);
+        }
+        elementData[size = newSize] = null;
```

Comment:
```suggestion
        elementData[ ??? ] = null;
```
let's decrement `size` and use it to make the previously last element `null`

Reply:
done